Migrate hidden block types (block manager data) to new preferences packages

talldan opened the pull request:
## Description
This PR is the next in a series of PRs migrating our editor preferences to use the new preferences package.

It tackles the `hiddenBlockTypes` state in the post editor that's used to drive the block manager, and should enable the block manager to be used within other editors (if we can figure out a nice way to migrate the component).

While #39115 handles all the boolean 'feature' preferences that are easy to migrate, the `hiddenBlockTypes` state has its own reducer, so I think it makes sense to have a separate PR to give this more scrutiny.

There are probably a few different ways to do this, but I've chosen to keep the code roughly equivalent, basically moving the `hiddenBlockType` reducer logic into the actions.

----

### An alternative I considered

An alternative could be to have an individual preference in the preferences store for each block and use a compound key, something like:
```js
export const hideBlockType = ( blockName ) => ( { registry } ) => {
	registry
		.dispatch( preferencesStore )
		.set( 'core/edit-post', `hiddenBlockTypes/${ blockName }`, true );
};

export const showBlockType = ( blockName ) => ( { registry } ) => {
	registry
		.dispatch( preferencesStore )
		.unset( 'core/edit-post', hiddenBlockTypes/${ blockName }` );
};
```

This would simplify the actions, but where there's a need to get the full list of hidden block types, that becomes harder. The selector would need to filter the preferences down to only those with the 'hiddenBlockTypes' prefix. That might be performance intensive, given selectors are generally called more often than actions.

## Testing Instructions
1. Delete your WP localstorage data
2. Using `trunk` open the post editor and dismiss the welcome guide
3. Open the preferences modal, and in the 'Blocks' tab, hide some of the blocks.
4. Check out this branch and rebuild
5. Reload the post editor
6. Open the preferences modal again and go back to the 'Blocks' tab. The blocks toggled in step 3, should still be the same. (this shows the local storage data migration worked)
7. Toggle a few more blocks
8. Reload
9. Open the preferences modal again and go back to the 'Blocks' tab. The preferences from step 7 (and step 3) should still be the same (the changes are still persisted)

talldan committed:
Use preferences store for hiddenBlockTypes

talldan committed:
Add unit tests

talldan committed:
Remove reducer

talldan commented on packages/edit-post/src/store/selectors.js:
```diff
@@ -105,11 +105,20 @@ export function getPreferences( state ) {
+			);
+		}
+
+		const preferences = getPreferences( state );
```
I need to do a bit more work on back compat here and make sure `getPreferences` has any migrated preferences merged into it.

noisysocks replied:
Would we then deprecate `getPreference`?

talldan replied:
Once everything is migrated I think it can be deprecated, yep 😄

noisysocks commented:
Looking good. I agree with this approach as opposed to the alternative you described.

I think that:

- we should deprecate the `getPreference` selector.
- the `hideBlockTypes` and `showBlockTypes` actions can stay as they add a slight bit of value by keeping the preference de-duped.
- we should add a `getHiddenBlockTypes` or `isBlockTypeHidden` selector to round out the API. If I were a plugin developer and I saw an `hideBlockTypes` action then I'd expect an analogous selector that I could call.
- we should add unit tests for the selectors and actions. It's easy to mock this stuff now that everything is thunkified.

Fix the tests and I'll review properly 😀

talldan committed:
Add merged prefs to `getPreferences`

talldan committed:
Add white space to test file

talldan committed:
Add tests for hiding and showing block types

talldan committed:
Add getHiddenBlockTypes selector

talldan committed:
Add tests for remaining preferences actions

talldan commented on packages/edit-post/src/store/test/actions.js:
```diff
@@ -129,4 +136,158 @@ describe( 'actions', () => {
+		} );
+	} );
+
+	describe( '__experimentalUpdateLocalAutosaveInterval', () => {
```
I've added tests for all the other preferences to bring some confidence that the changes to `getPreferences` and `getPreference` were correct. Plus they'll help when I come to migrate the other preferences.

talldan commented:
> we should add a getHiddenBlockTypes or isBlockTypeHidden selector to round out the API. If I were a plugin developer and I saw an hideBlockTypes action then I'd expect an analogous selector that I could call.

Agree with that, I immediately went looking for it. 😄 

It does mean a selector that doesn't really do much other than get something from the preferences store, but that's probably ok. This is also going to set the pattern for every other editor that adds the block manager, which is the only other thing that concerns me.

talldan commented:
I believe I've addressed all the feedback.

Registry selectors are a little tricky to test, so I've mostly tested them within the action tests. I think it's enough coverage

noisysocks commented on docs/reference-guides/data/data-core-edit-post.md:
```diff
@@ -81,6 +81,15 @@ _Returns_
+### getHiddenBlockTypes
+
+Registry selector that gets the hidden block types from the preferences
+store.
```
I think the descriptions for the selector and two actions are a little too descriptive of the implementation. A plugin author probably doesn't care how it's implemented, just what happens when e.g. `hideBlockTypes` is dispatched, i.e. the block is hidden from the inserters.

noisysocks reviewed (approved):
Fantastic! 👍

talldan committed:
Refine documentation
